finamp: init at 0.9.8-beta

DSeeLP opened the pull request:
## Description of changes
[Finamp](https://github.com/jmshrv/finamp)
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

DSeeLP committed:
finamp: init at 0.9.8-beta

DSeeLP committed:
maintainers: add dseelp

nicolas-goudry commented on pkgs/by-name/fi/finamp/package.nix:
```diff
@@ -0,0 +1,69 @@
+    license = licenses.mpl20;
+    platforms = platforms.linux;
+    mainProgram = "finamp";
+    maintainers = with lib.maintainers; [ dseelp ];
```
This line is already scoped with `lib`, either:

* use `with maintainers` (preferred)
* keep it as is but unscope other attributes (ie. `license` and `platforms`)
* unscope everything

DSeeLP replied:
I did the 1. thing

nicolas-goudry commented on pkgs/by-name/fi/finamp/package.nix:
```diff
@@ -0,0 +1,69 @@
+
+  desktopItem = makeDesktopItem {
+    name = "Finamp";
+    exec = "@out@/bin/finamp";
```
```suggestion
    exec = "finamp";
```

I think you can just use `"finamp"` here, since upon installation it will be in the PATH.

This would also remove line 45 (`substituteInPlace`).

DSeeLP replied:
I changed it and removed the call to `substituteInPlace`

nicolas-goudry commented on pkgs/by-name/fi/finamp/package.nix:
```diff
@@ -0,0 +1,69 @@
+  };
+
+  meta = with lib; {
+    description = "A Jellyfin music client for mobile and desktop";
```
Just a suggestion, but since NixOS isn’t on mobile (afaik), I would keep the description concise and simple with something like « A Jellyfin music client ». Up to you!

If you change it, also change the `genericName` attribute of `makeDesktopItem`.

DSeeLP replied:
I changed it to "An open source Jellyfin music player" and added it to the `makeDesktopItem` `comment` field and I changed the `genericName` field to `Music Player` like it is done upstream.

nicolas-goudry commented:
Result of `nixpkgs-review pr 315139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>finamp</li>
    <li>finamp.debug</li>
    <li>finamp.pubcache</li>
  </ul>
</details>

nicolas-goudry commented:
I was able to run the program and connect to my Jellyfin server. I don’t have any music on it so I couldn’t test the actual media playback.

DSeeLP commented:
Media playback works for me, but there is currently no built in volume control.

nicolas-goudry commented on maintainers/maintainer-list.nix:
```diff
@@ -5532,6 +5532,11 @@
+  dseelp = {
+    name = "dsee";
+    github = "DSeeLP";
+    githubId = "46624152";
```
I didn’t catch that earlier, the githubId must be an integer.

DSeeLP replied:
Fixed

nicolas-goudry commented:
I was able to add some music to my Jellyfin server and can confirm the playback is working.

This may be related to the software being in beta, but there are a few graphical issues in the app.

DSeeLP commented:
Result of `nixpkgs-review pr 315139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>finamp</li>
    <li>finamp.debug</li>
    <li>finamp.pubcache</li>
  </ul>
</details>

DSeeLP commented:
Updated to 0.9.8-beta which includes a few bug fixes

nicolas-goudry commented on pkgs/by-name/fi/finamp/package.nix:
```diff
@@ -0,0 +1,69 @@
+    startupWMClass = "finamp";
+    comment = "An open source Jellyfin music player";
+    categories = [
+      "Audio"
```
Just noticed this while building:

```
/nix/store/5ddf158z5769lbw9biq97npk0sxfd061-Finamp.desktop/share/applications/Finamp.desktop: error: (will be fatal in the future): value item "Audio" in key "Categories" in group "Desktop Entry" requires another category to be present among the following categories: AudioVideo
```

Adding `AudioVideo` to the categories should fix it.

DSeeLP replied:
I added `AudioVideo` to the categories.
[Flutter first tries `/etc/os-release` and falls back to `/usr/lib/os-release` if it doesn't exist.](https://github.com/flutter/flutter/blob/5187cab7bdd434ca74abb45895d17e9fa553678a/packages/flutter_tools/lib/src/base/os.dart#L321)
I noticed that during build `/etc` only contains `group`, `hosts` and `passwd`.

nicolas-goudry replied:
Oh yes, that may be because builds are sandboxed.

nicolas-goudry reviewed (changes requested):
Most likely not an issue but there is the following error when building:

```
Failed obtaining PRETTY_NAME for Linux: PathNotFoundException: Cannot open file, path = '/usr/lib/os-release' (OS Error: No such file or directory, errno = 2)
```

Don’t know why the build is using `/usr/lib/os-release`, I always use `/etc/os-release` (which exists on NixOS).

Would be great to patch this upstream or in the package! But not mandatory IMO, since the package builds and works fine.

pyrox0 commented:
This depends on the `xdg-user-dir` tool, which I found out because my system configuration does not include this by default. Please add `xdg-user-dirs`(note the extra s on the end) to `buildInputs` so that the package runs properly if this is not installed seperately!

I've confirmed this fix works by getting a nix shell with this tool, and it works properly.

DSeeLP commented:
I added `AudioVideo` to the categories and added `xdg-user-dirs` to the `buildInputs`.

pyrox0 commented:
LGTM!

Titaniumtown reviewed (approved):
LGTM

Titaniumtown commented on pkgs/by-name/fi/finamp/package.nix:
```diff
@@ -0,0 +1,71 @@
+  pubspecLock = lib.importJSON ./pubspec.lock.json;
+
+  nativeBuildInputs = [ patchelf ];
+  buildInputs = [ mpv xdg-user-dirs ];
```
`xdg-user-dirs` also needs to be a run-time dependency.

DSeeLP replied:
I added it to `PATH` using `extraWrapProgramArgs` and removed it from `buildInputs`.

Titaniumtown replied:
Ok great. Works perfectly now @DSeeLP 
Good work!

Aleksanaa commented on pkgs/by-name/fi/finamp/package.nix:
```diff
@@ -0,0 +1,71 @@
+  };
+
+  meta = with lib; {
+    description = "An open source Jellyfin music player";
```
```suggestion
    description = "Open source Jellyfin music player";
```

DSeeLP replied:
Fixed

Aleksanaa commented on pkgs/by-name/fi/finamp/package.nix:
```diff
@@ -0,0 +1,71 @@
+    cp $src/assets/icon/icon_foreground.svg $d/finamp.svg
+
+    mkdir -p $out/share/applications
+    cp $desktopItem/share/applications/*.desktop $out/share/applications
```
Use `desktopItems` with `copyDesktopItems` hook instead.

DSeeLP replied:
Fixed

Aleksanaa commented on pkgs/by-name/fi/finamp/package.nix:
```diff
@@ -0,0 +1,71 @@
+    ];
+  };
+
+  meta = with lib; {
```
Better to avoid `with lib;`

nicolas-goudry replied:
That’s not the first time I see another reviewer say this. What is the rationale? At such a low level, it doesn’t introduce much hassle to understand the code. Would it be placed at a higher level I would agree with you, but I don’t understand why it would be an issue here.

Aleksanaa replied:
See https://github.com/NixOS/nixpkgs/issues/208242

nicolas-goudry replied:
Thanks for pointing to this issue. As I said earlier, I do understand and do agree that using `with` at a high level is clearly misleading and makes the code much harder to read, even more when multiple `with` are used.

However, here I can only agree with SuperSandro and eclairevoyant:

> The problem is not when with is used in meta where it is very local and only covering a few lines. I am completely against changing those treewide.

> The nix.dev antipatterns guide warns against using with at the top of a Nix file - a reasonable suggestion, but does not mention anything about using with in the very predictable/standardised spot of only in meta.

I believe the `with` usages on `meta` attributes are legit since they save a few keystrokes while keeping the code readable, even for beginners. IMO this should not be changed.

DSeeLP replied:
I will change it to match the manual at https://nixos.org/manual/nixpkgs/unstable/#chap-meta

DSeeLP commented:
I changed the meta section to match the [nixpkgs manual](https://nixos.org/manual/nixpkgs/unstable/#chap-meta) and adjusted the package description.
I removed `xdg-user-dirs` from `buildInputs` and added it to `PATH` using `extraWrapProgramArgs` and also switched to using the `copyDesktopItems` hook instead of copying the desktop file in `postInstall`.

DSeeLP commented:
3 packages built:
finamp finamp.debug finamp.pubcache

Result of `nixpkgs-review pr 315139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>finamp</li>
    <li>finamp.debug</li>
    <li>finamp.pubcache</li>
  </ul>
</details>

Guanran928 reviewed (approved):
Works on `x86_64-linux`.

Saw some flickering, but it disappeared later.

https://github.com/NixOS/nixpkgs/assets/68757440/734fa766-f7c1-4448-b455-7c8e2ac721be

Aleksanaa commented on pkgs/by-name/fi/finamp/package.nix:
```diff
@@ -0,0 +1,73 @@
+let
+  version = "0.9.8-beta";
+in
+flutter.buildFlutterApplication {
```
The flutter version should pin to upstream version. See https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/dart.section.md#flutter-applications-ssec-dart-flutter.

```suggestion
flutter322.buildFlutterApplication {
```

You can find the version in it's workflow run log (of the `0.9.8-beta` commit): https://github.com/jmshrv/finamp/actions/runs/9504242145/job/26196438043#step:5:38

DSeeLP replied:
Fixed

Aleksanaa commented on pkgs/by-name/fi/finamp/package.nix:
```diff
@@ -0,0 +1,73 @@
+{ lib
+, flutter
+, mpv
```
```suggestion
, mpv-unwrapped
```

Not so sure, but usually we don't use wrapped ones if we need the library

DSeeLP replied:
Fixed

Aleksanaa commented on pkgs/by-name/fi/finamp/package.nix:
```diff
@@ -0,0 +1,73 @@
+  postInstall = ''
+    d=$out/share/icons/hicolor/scalable/apps
+    mkdir -p $d
+    cp $src/assets/icon/icon_foreground.svg $d/finamp.svg
```
```suggestion
    install -Dm644 $src/assets/icon/icon_foreground.svg $d/finamp.svg
```

No need to `mkdir`.

DSeeLP replied:
Fixed

DSeeLP commented:
I applied all suggestions from @Aleksanaa to use `install` instead of `mkdir` and `cp` afterwards and to use the unwrapped of mpv version and to specify the flutter release.
@Guanran928 the flickering also happens to me. It also happens on Ubuntu 24.04 with the release binaries. I reported it https://github.com/jmshrv/finamp/issues/801.

DSeeLP commented:
3 packages built:
finamp finamp.debug finamp.pubcache

Result of `nixpkgs-review pr 315139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>finamp</li>
    <li>finamp.debug</li>
    <li>finamp.pubcache</li>
  </ul>
</details>
